File: ios/ForexZoneApp/ForexZoneApp/Models/PairConfig.swift
import Foundation

/// Per-pair-per-timeframe configuration from the MCP server
struct PairConfig: Identifiable, Codable, Hashable {
    var id: String { "\(Instrument)_\(ZoneGranularity)" }

    let Instrument: String
    let ZoneGranularity: String
    let TrendGranularity: String
    let Enabled: Bool
    let EmailEnabled: Bool
    let MinBaseLength: Int?
    let MaxBaseLength: Int?
    let MinLegInToBaseRangeRatio: Double?
    let MinLegOutToBaseRangeRatio: Double?
    let SwingLookback: Int?
    let TrendCandleCount: Int?
    let MinSwingPoints: Int?
    let ConfigVersion: Int
    let UpdatedAtUtc: String?

    // Status fields (from joined response)
    let LastProcessedUtc: String?
    let ZoneCount: Int?
    let Trend: String?

    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }

    static func == (lhs: PairConfig, rhs: PairConfig) -> Bool {
        lhs.id == rhs.id
    }
}

struct PairConfigListResponse: Codable {
    let TotalConfigs: Int
    let Configs: [PairConfig]
}

struct PairConfigUpdateResponse: Codable {
    let Status: String?
    let Instrument: String?
    let Granularity: String?
    let ConfigVersion: Int?
    let Error: String?
}

/// Status-only response
struct PairStatusResponse: Codable {
    let Instrument: String?
    let Granularity: String?
    let LastProcessedUtc: String?
    let ConfigVersionProcessed: Int?
    let ZoneCount: Int?
    let Trend: String?
    let Error: String?
}

File: src/ForexZoneAnalyzer.Worker/Configuration/PairConfig.cs
namespace ForexZoneAnalyzer.Worker.Configuration;

/// <summary>
/// Per-instrument-per-timeframe configuration for zone and trend detection.
/// Stored in Azure Table Storage (pairconfigs table).
/// PartitionKey = instrument, RowKey = zone granularity.
/// </summary>
public class PairConfig
{
    public string Instrument { get; set; } = "";
    public string ZoneGranularity { get; set; } = "";
    public string TrendGranularity { get; set; } = "";

    // Processing control
    public bool Enabled { get; set; }
    public bool EmailEnabled { get; set; }

    // Zone detection parameters
    public int MinBaseLength { get; set; } = 1;
    public int MaxBaseLength { get; set; } = 6;
    public double MinLegInToBaseRangeRatio { get; set; } = 1.0;
    public double MinLegOutToBaseRangeRatio { get; set; } = 1.0;

    // Trend detection parameters
    public int SwingLookback { get; set; } = 3;
    public int TrendCandleCount { get; set; } = 60;
    public int MinSwingPoints { get; set; } = 2;

    // Versioning (incremented on config change; triggers zone refresh)
    public int ConfigVersion { get; set; } = 1;
    public DateTime UpdatedAtUtc { get; set; } = DateTime.UtcNow;

    public ZoneAnalyzer.PatternAnalysis.ZoneConfiguration ToZoneConfiguration() => new()
    {
        MinBaseLength = MinBaseLength,
        MaxBaseLength = MaxBaseLength,
        MinLegInToBaseRangeRatio = MinLegInToBaseRangeRatio,
        MinLegOutToBaseRangeRatio = MinLegOutToBaseRangeRatio
    };

    public ZoneAnalyzer.PatternAnalysis.TrendConfiguration ToTrendConfiguration() => new()
    {
        SwingLookback = SwingLookback,
        TrendCandleCount = TrendCandleCount,
        MinSwingPoints = MinSwingPoints
    };
}
